electrum/wallet.py

```diff
@@ -1837,6 +1837,37 @@ def get_new_sweep_address_for_channel(self) -> str:
         assert is_address(selected_addr), f"not valid bitcoin address: {selected_addr}"
         return selected_addr
 
+    def should_keep_reserve_utxo(
+            self,
+            tx_inputs: List[PartialTxInput],
+            tx_outputs: List[PartialTxOutput],
+            is_anchor_channel_opening: bool,
+    ) -> bool:
+        channels_need_reserve = self.lnworker and any(chan.has_anchors() and not chan.is_redeemed() for chan in self.lnworker.channels.values())
+        # note: is_anchor_channel_opening is used in unit tests, without lnworker
+        is_reserve_needed = is_anchor_channel_opening or channels_need_reserve
+        if not is_reserve_needed:
+            return False
+
+        coins_in_wallet = self.get_spendable_coins(nonlocal_only=False, confirmed_only=False)
+        prevout_coins_in_wallet = set(c.prevout for c in coins_in_wallet)
+        amount_in_wallet = sum(c.value_sats() for c in coins_in_wallet)
+
+        amount_consumed = sum(c.value_sats() for c in tx_inputs if c.prevout in prevout_coins_in_wallet)
+        amount_retained = sum(o.value for o in tx_outputs if self.is_mine(o.address))
+        to_be_spent_sat = amount_consumed - amount_retained
+
+        assert amount_in_wallet - to_be_spent_sat >= 0
```

Comment:
This assert can fail e.g. due to multi-threading races. Consider that
- thread1: make_unsigned_transaction gets called,
- thread2: spends some coins or we learn of a new outgoing tx from the network,
- thread1: make_unsigned_transaction reaches this assert but `coins_in_wallet` is now of much lesser value than at step1

Not sure how to practically fix. Just wanted to point out the conceptual issue.
If you added a code comment about this here with a fixme, I would say that's enough.

Reply:
sorry, didn't see this comment, you added it while I was merging. 
I guess the proper way to fix it is to add a proper locking mechanism on everything that affects available coins.
The scope of such a mechanism is potentially large.